test_runner/regress/test_remote_storage.py

```diff
@@ -717,6 +717,59 @@ def test_empty_branch_remote_storage_upload_on_restart(
     ), f"New branch should have been reuploaded on pageserver restart to the remote storage path '{new_branch_on_remote_storage}'"
 
 
+# Test creates >1000 timelines and upload them to the remote storage.
+# AWS S3 does not return more than 1000 items and starts paginating, ensure that pageserver paginates correctly.
+@pytest.mark.parametrize("remote_storage_kind", [RemoteStorageKind.MOCK_S3])
+def test_thousands_of_branches(
```

Comment:
The test is quite high level. To test the issue it is enough to have way less complicated test which is limited to scope of our s3 client. I wonder what will be the difference in runtime? 

We already have infra in place to run tests on real s3. So this can be an integration test under libs/remote_storage/tests. WDYT?

Reply:
My biggest concern are periodic real-s3 e2e test breakages, with extra 1000 something entities to upload — will it affect other test runs, if every PR starts to do that?

I see that there's `max-keys` url param in https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_RequestSyntax that could be used to reduce the load, but the question is how to add that into s3 requests.

Reply:
Hm, thats a good find.

We have somewhat similar `UnreliableWrapper` for testing purposes.

Options I see: 
1. Still go try implement the test in rust, it will be simpler to pass parameter here (cfg(test) or something)
2. Add new types of testing remote storage impls and allow to decode them from config when built with testing flags.

Reply:
Took me a while to create on of the first Rust e2e tests, but here is the alternative: https://github.com/neondatabase/neon/pull/3898

It does run faster, but the amount of code and various checks is bigger, IMO, it's harder to run this test locally too.
On the good side, it made me to pay attention into the `list` command, I think it does not work for `S3` impl normally, and since we don't use it, should we actually remove it?

WDYT on both the PR and the removal proposition?

Reply:
https://github.com/neondatabase/neon/pull/3898 merged and rebased